va/va_test.go

```diff
@@ -1117,3 +1118,206 @@ func TestParseResults(t *testing.T) {
 	test.Assert(t, s.Unknown[0] == &expected, "Incorrect record returned")
 	test.AssertNotError(t, err, "no error should be returned")
 }
+
+func TestAvailableAddresses(t *testing.T) {
+	v6a := net.ParseIP("::1")
+	v6b := net.ParseIP("2001:db8::2:1") // 2001:DB8 is reserved for docs (RFC 3849)
+	v4a := net.ParseIP("127.0.0.1")
+	v4b := net.ParseIP("192.0.2.1") // 192.0.2.0/24 is reserved for docs (RFC 5737)
+
+	testcases := []struct {
+		input core.ValidationRecord
+		v4    []net.IP
+		v6    []net.IP
+	}{
+		// An empty validation record
+		{
+			core.ValidationRecord{},
+			[]net.IP{},
+			[]net.IP{},
+		},
+		// A validation record with one IPv4 address
+		{
+			core.ValidationRecord{
+				AddressesResolved: []net.IP{v4a},
+			},
+			[]net.IP{v4a},
+			[]net.IP{},
+		},
+		// A dual homed record with an IPv4 and IPv6 address
+		{
+			core.ValidationRecord{
+				AddressesResolved: []net.IP{v4a, v6a},
+			},
+			[]net.IP{v4a},
+			[]net.IP{v6a},
+		},
+		// The same as above but with the v4/v6 order flipped
+		{
+			core.ValidationRecord{
+				AddressesResolved: []net.IP{v6a, v4a},
+			},
+			[]net.IP{v4a},
+			[]net.IP{v6a},
+		},
+		// A validation record with just IPv6 addresses
+		{
+			core.ValidationRecord{
+				AddressesResolved: []net.IP{v6a, v6b},
+			},
+			[]net.IP{},
+			[]net.IP{v6a, v6b},
+		},
+		// A validation record with interleaved IPv4/IPv6 records
+		{
+			core.ValidationRecord{
+				AddressesResolved: []net.IP{v6a, v4a, v6b, v4b},
+			},
+			[]net.IP{v4a, v4b},
+			[]net.IP{v6a, v6b},
+		},
+	}
+
+	for _, tc := range testcases {
+		// Split the input record into v4/v6 addresses
+		v4result, v6result := availableAddresses(tc.input)
+
+		// Test that we got the right number of v4 results
+		test.Assert(t, len(tc.v4) == len(v4result),
+			fmt.Sprintf("Wrong # of IPv4 results: expected %d, got %d", len(tc.v4), len(v4result)))
+
+		// Check that all of the v4 results match expected values
+		for i, v4addr := range tc.v4 {
+			test.Assert(t, v4addr.String() == v4result[i].String(),
+				fmt.Sprintf("Wrong v4 result index %d: expected %q got %q", i, v4addr.String(), v4result[i].String()))
+		}
+
+		// Test that we got the right number of v6 results
+		test.Assert(t, len(tc.v6) == len(v6result),
+			fmt.Sprintf("Wrong # of IPv6 results: expected %d, got %d", len(tc.v6), len(v6result)))
+
+		// Check that all of the v6 results match expected values
+		for i, v6addr := range tc.v6 {
+			test.Assert(t, v6addr.String() == v6result[i].String(),
+				fmt.Sprintf("Wrong v6 result index %d: expected %q got %q", i, v6addr.String(), v6result[i].String()))
+		}
+	}
+}
+
+func TestFallbackDialer(t *testing.T) {
+	// Create a test VA
+	va, _, _ := setup()
+
+	// Create a new challenge to use for the httpSrv
+	chall := core.HTTPChallenge01()
+	setChallengeToken(&chall, core.NewToken())
```

Comment:
I tried - in this case the test fails without it. `setChallengeToken` sets the `ProvidedKeyAuthorization` to match the token value and `core.HTTPChallenge01()` alone doesn't.